### ISSUE TITLE
transceiver.stop() shouldn't block in-progress replaceTrack from succeeding

### ISSUE THREAD

**Issue by jan-ivar** [w3c/webrtc-pc#2106]
Most WebRTC methods' final queued task say:

   1. *If connection's [[IsClosed]] slot is true, then abort these steps.*

But [sender.replaceTrack](http://w3c.github.io/webrtc-pc/#dom-rtcrtpsender-replacetrack)'s says:

   1. *If transceiver's [[Stopped]] slot is true, abort these steps.*

That may seem like the sender-scoped equivalent, but has the side-effect that `transceiver.stop()` causes an already in-progress `replaceTrack` to never resolve. That seems unintended.

I propose we make replaceTrack like the other methods.

One question then is should the promise resolve or reject with an error? I think resolving is fine, since it's a glorified setter at that point; no harm will come of it.

**Comment by youennf** [comment]
I know this is an existing practice for RTCPeerConnection but it does not seem like a good pattern for an API to return a promise and never reject/resolve it. That may leave code to be hanging forever due to some timing issues.
I would just remove step 6.4.4.1.

**Comment by jan-ivar** [comment]
@youennf I believe you're arguing for less blocking. This issue and PR result in less blocking, so I'm not hearing an argument against merging this.

We should be consistent, and this PR merely aligns with our existing pattern (removing unintended blocking). This isn't theoretical, as we're fixing our Firefox code right now, and would like to merge this spec support to avoid blocking on `stop()`. I suggest opening a separate issue to challenge the overall pattern.

> That may leave code to be hanging forever due to some timing issues.

FWIW I think it's inaccurate to characterize the absence of a callback as "hanging".

**Comment by aboba** [comment]
Agree that the promise should be rejected or resolved. However, we don't seem to have been consistent with respect to this throughout the specification. 

For example, several of the "If connection's [[IsClosed]] slot is true, then abort these steps." instances would appear to result in a promise never being rejected or resolved.  An example is `addIceCandidate`.  Was this what we really intended? 

Note that Step 6 sub-step 1 already tests if `transceiver` has been stopped, and rejects the promise if it has. Would it make more sense for `replaceTrack` to reject the promise if it discovers a closed `RTCPeerConnection` or `transceiver` later?

**Comment by jan-ivar** [comment]
> Agree that the promise should be rejected or resolved.

That's not actually a thing in JS.

> However, we don't seem to have been consistent with respect to this throughout the specification.
>
> For example, several of the "If connection's [[IsClosed]] slot is true, then abort these steps." instances would appear to result in a promise never being rejected or resolved. An example is `addIceCandidate`. Was this what we really intended?

I believe we're consistent: Outstanding Pomises are left pending on `pc.close()`. https://github.com/w3c/webrtc-pc/issues/150#issuecomment-152333247

**Comment by aboba** [comment]
Yes, we do appear to be consistent with respect to `pc.close().  Thanks for the link to the debate on this.

**Comment by youennf** [comment]
> I suggest opening a separate issue to challenge the overall pattern.

Sure, I understand that RTCPeerConnection has this behavior.

Other APIs I know of usually resolve or reject promises whenever possible, except in some specific cases, like the page is navigating for instance or the task is still pending.
In this particular case, there is no pending task, so the expectation is to either resolve or reject the promise.

Although this particular RTCPeerConnection is surprising, this issue is a separate issue.

> I believe we're consistent: Outstanding Pomises are left pending on `pc.close()`. [#150 (comment)](https://github.com/w3c/webrtc-pc/issues/150#issuecomment-152333247)

RTCRtpSender is not consistently doing that.
RTCRtpSender.setParameters and RTCRtpSender.getStats are not checking for stopped/closed before resolving/rejecting the promise.

Note also that the PR further ties RTCRtpSender with RTCPeerConnection, which is not great.
It does not align with the idea of being able to create objects like transceivers or data channels directly from an ice connection.

What is the benefit of the PR vs. removing step 6.4.4.1, or replacing step 6.4.4.1 by rejecting the promise with InvalidStateError?

**Comment by jan-ivar** [comment]
FWIW I don't think it's obvious how async operations should react to close/abort/navigation. Happy to look at other APIs, but most specs don't discuss navigation at all.

Without rehashing https://github.com/w3c/webrtc-pc/issues/150#issuecomment-152333247, it's worth noting "cancelable promises" were hotly debated at the time, with a potential "third rail" of propagation for cancellation solely thru `finally` statements.

That got scrapped. Instead, `fetch` adopted a simpler [abortController](https://developers.google.com/web/updates/2017/09/abortable-fetch) model that rejects with `AbortError`. This might count as new information if you want to re-open the issue.

I think we need practical reasons though. There's nothing inherently wrong with pending promises:
```js
(async () => {
  let id;
  button.onclick = () => clearTimeout(id);
  await new Promise(r => id = setTimeout(r, 10000));
  console.log("The sky is falling");
})();
```

```js
const id = setTimeout(() => console.log("No-one cared ever"), 10000);
button.onclick = () => clearTimeout(id);
```
Garbage collection cleans up both AFAIK.

I think we need to ask: Why would an app want renegotiation errors after having called `pc.close()`?

**Comment by jan-ivar** [comment]
> > I believe we're consistent: Outstanding Pomises are left pending on `pc.close()`.
> RTCRtpSender is not consistently doing that.

Sorry, I should have said consistent with the other methods using the [peer connection queue](http://w3c.github.io/webrtc-pc/#enqueue-an-operation).

> RTCRtpSender.setParameters and RTCRtpSender.getStats are not checking for stopped/closed before resolving/rejecting the promise.

*getStats* is specifically allowed on closed peer connections.

*setParameters* is not enqueued, and is not allowed to fail with `InvalidStateError`, so I'd assume it succeeds, which seems benign enough.

> Note also that the PR further ties RTCRtpSender with RTCPeerConnection, which is not great.

That ship has sailed I think: *replaceTrack* is enqueued on the peer connection queue, for reasons.

> It does not align with the idea of being able to create objects like transceivers or data channels directly from an ice connection.

You can't do that today. We would need to spec that, but I think we're getting ahead of ourselves.

> What is the benefit of the PR vs. removing step 6.4.4.1, or replacing step 6.4.4.1 by rejecting the promise with InvalidStateError?

Consistency with the other enqueued methods. Having `transceiver.stop()` cause outstanding *replaceTrack* promises to reject with `InvalidStateError` would be novel. I'm not hearing support for that.

**Comment by youennf** [comment]
> > RTCRtpSender.setParameters and RTCRtpSender.getStats are not checking for stopped/closed before resolving/rejecting the promise.
> 
> _getStats_ is specifically allowed on closed peer connections.

Right, so I would not assume anybody to write code in a way that handle the case of some WebRTC methods returning promises that will never reject/resolve.

> _setParameters_ is not enqueued, and is not allowed to fail with `InvalidStateError`, so I'd assume it succeeds, which seems benign enough.

So, we are asking web developers to know that setParameters is not enqueuing and replaceTrack is enqueuing, so that they properly handle the fact that a promise may never be settled in some cases for replaceTrack?

> Consistency with the other enqueued methods. Having `transceiver.stop()` cause outstanding _replaceTrack_ promises to reject with `InvalidStateError` would be novel. I'm not hearing support for that.

I expect more consistency between the methods of a given class than between methods of different classes.

As I said, I am fine with either resolve or reject.
Rejecting would be novel but web apps need to handle the InvalidStateError rejection anyway so this might not be a big deal in practice.

Resolving might indeed be less risky.
The additional bonus of resolving are that it reduces the code to write and the time to read the spec.

> I think we need to ask: Why would an app want renegotiation errors after having called `pc.close()`?

If you look at streams API which heavily relies on promises, why would an app want to get read promise errors after having called stream.abort?
An example might be that one part of the code of an app might do read.
Another part of the code might do abort but does not know about the one waiting to read or might forget to notify it.

For WebRTC, one purpose might be clean-up.
replaceTrack promise.finally might be used to know when to pause the audio context fueling an outgoing audio track, something like ref counting on the use of the audio context to decide whether to pause/resume it.
For pc, a repeating timer could be created during renegotiation and cleared at end of negotiation.
Everything would be fine until one day the app is updated to call pc.close to follow good patterns, or fix some leak issues. Now, the repeating timer might in some code paths never be cleared...
